### ISSUE TITLE
Sparse arrays

### ISSUE THREAD

**Issue by mrocklin** [pydata/xarray#1375]
I would like to have an XArray that has scipy.sparse arrays rather than numpy arrays.  Is this in scope?  

What would need to happen within XArray to support this?

**Comment by shoyer** [comment]
Currently, we define most of our operations on duck arrays in https://github.com/pydata/xarray/blob/master/xarray/core/duck_array_ops.py

There are a few other hacks throughout the codebase, which can find by searching for "dask_array_type": https://github.com/pydata/xarray/search?p=1&q=dask_array_type&type=&utf8=%E2%9C%93

It's pretty crude, but basically this would need to be extended to implement many of these methods on for sparse arrays, too. Ideally we would define xarray's adapter logic into more cleanly separated submodules, perhaps using multiple dispatch. Even better, we would make this public API, so you can write something like `xarray.register_data_type(MySparseArray)` to register a type as valid for xarray's `.data` attribute.

It looks like `__array_ufunc__` will actually finally land in NumPy 1.13, which might make this easier.

See also https://github.com/pydata/xarray/pull/1118

**Comment by rabernat** [comment]
👍  to the scipy.sparse array suggestion

[While we are discussing supporting other array types, we should keep gpu arrays on the radar:
https://documen.tician.de/pycuda/array.html]

**Comment by benbovy** [comment]
Although I don't know much about [SciDB](http://scidb-py.readthedocs.io/en/stable/), it seems to be another possible application for  `xarray.register_data_type`.

**Comment by mrocklin** [comment]
Here is a brief attempt at a multi-dimensional sparse array: https://github.com/mrocklin/sparse

It depends on numpy and scipy.sparse and, with the exception of a bit of in-memory data movement and copies, should run at scipy speeds (though I haven't done any benchmarking). 

@rabernat do you have an application that we could use to drive this?

**Comment by rabernat** [comment]
> @rabernat do you have an application that we could use to drive this?

Nothing comes to mind quickly. My data are unfortunately quite dense! 😜

**Comment by olgabot** [comment]
In case you're still looking for an application, [gene expression from single cells](https://github.com/olgabot/macosko2015) (see `data/00_original/GSM162679$i_P14Retina_$j.digital_expression.txt.gz`) is very sparse due to high gene dropout. The shape is `expression.shape (49300, 24760)` and it's mostly zeros or nans. A plain csv from this data was 2.5 gigs, which gzipped to 300 megs.

[Here](https://github.com/olgabot/macosko2015/blob/master/notebooks/05_combine_retina_data.ipynb) is an example of using `xarray` to combine these files but my kernel keeps dying when I do `ds.to_netcdf()` :( 

Hope this is a good example for sparse arrays!

**Comment by rth** [comment]
> do you have an application that we could use to drive this?

Other examples where labeled sparse arrays would be useful are,
  * [one-hot encoding](https://rasbt.github.io/mlxtend/user_guide/preprocessing/one-hot_encoding/) that are widely used in machine learning. 
  * [tokenizing textual data](http://scikit-learn.org/stable/modules/feature_extraction.html#common-vectorizer-usage) produces large sparse matrices where the column labels correspond to the vocabulary, while row labels correspond to document ids. Here is a minimal example using scikit-learn,
     ```py
     import os.path
     from sklearn.datasets import fetch_20newsgroups
     from sklearn.feature_extraction.text import CountVectorizer

     ds = fetch_20newsgroups()
     vect = CountVectorizer()
     X = vect.fit_transform(ds.data)
     print(X)  # Extracted tokens
     # Returns:
     # <11314x130107 sparse matrix of type '<class 'numpy.int64'>'
     #	with 1787565 stored elements in Compressed Sparse Row format>

     column_labels = vect.get_feature_names()
     print(np.asarray(column_labels))
     # Returns:
     # array(['00', '000', '0000', ..., 'íålittin', 'ñaustin', 'ýé'],   dtype='<U180')

     row_labels = [int(os.path.split(el)[1]) for el in ds.filenames]
     print(np.asarray(row_labels))
     # Returns:
     # array([102994,  51861,  51879, ...,  60695,  38319, 104440])
     ```

**Comment by rabernat** [comment]
Sparse Xarray DataArrays would be useful for the linear regridding operations discussed in JiaweiZhuang/xESMF#3.

**Comment by lbybee** [comment]
I'm interested to see if there have been any developments on this.  I currently have an application where I'm working with multiple dask arrays, some of which are sparse (text data).  It'd be worth my time to move my project to xarray, so I'm be interested in contributing something here if there is a need.

**Comment by Hoeze** [comment]
I'd know a project which could make perfect use of xarray, if it would support sparse tensors:
https://github.com/theislab/anndata

Currently I have to convert my data between xarray and anndata every time I'm loading count data.
Therefore, I'd love having sparse data support in xarray :)

**Comment by shoyer** [comment]
See also: https://github.com/pydata/xarray/issues/1938

The major challenge now is the dispatching mechanism, which hopefully http://www.numpy.org/neps/nep-0018-array-function-protocol.html will solve.

**Comment by Hoeze** [comment]
Would it be an option to use dask's sparse support?
http://dask.pydata.org/en/latest/array-sparse.html
This way xarray could let dask do the main work.

Currently I load everything into a dask array by hand and pass this dask array to xarray.
This works pretty good.

**Comment by Hoeze** [comment]
How should these sparse arrays get stored in NetCDF4?
I know that NetCDF4 has some conventions how to store sparse data, but do we have to implement our own conversion mechanisms for each sparse type?

**Comment by shoyer** [comment]
> Would it be an option to use dask's sparse support?
http://dask.pydata.org/en/latest/array-sparse.html
This way xarray could let dask do the main work.

In principle this would work, though I would prefer to support it directly in xarray, too.

> I know that NetCDF4 has some conventions how to store sparse data, but do we have to implement our own conversion mechanisms for each sparse type?

Yes, we would need to implement a convention for handling sparse array data.

**Comment by rabernat** [comment]
Given the recent improvements in numpy duck array typing, how close are we to being able to just wrap a pydata/sparse array in an xarray Dataset?

**Comment by shoyer** [comment]
It will need some experimentation, but I think things should be pretty close after NumPy 1.17 is released. Potentially it could be as easy as adjusting the rules xarray uses for casting in `xarray.core.variable.as_compatible_data`.

**Comment by rabernat** [comment]
If someone who is good at numpy shows up at our sprint tomorrow, this could be a good issue try out.

**Comment by mrocklin** [comment]
@rgommers might be able to recommend someone

**Comment by rgommers** [comment]
I haven't talked to anyone at SciPy'19 yet who was interested in sparse arrays, but I'll keep an eye out today.

And yes, this is a fun issue to work on and would be really nice to have!

**Comment by rabernat** [comment]
I personally use the new sparse project for my day-to-day research. I am motivated on this, but I probably won't have time today to dive deep on this.

Maybe CuPy would be more exciting.

**Comment by mrocklin** [comment]
@nvictus has been working on this at #3117

**Comment by fjanoos** [comment]
Wondering what the status on this is ? Is there a branch with this functionality implemented - would love to give it a spin !

**Comment by shoyer** [comment]
This is working now on the `master` branch!

Once we get a few more kinks worked out, it will be in the next release.

I've started another issue for discussing how xarray could integrate sparse arrays better into its API: https://github.com/pydata/xarray/issues/3213

**Comment by fjanoos** [comment]
@shoyer 
Is there documentation for using sparse arrays ? Could you point me to some example code ?

**Comment by dcherian** [comment]
@fjanoos there isn't any formal documentation yet but you can look at test_sparse.py for examples. That file will also tell you what works and doesn't work currently.
